org.spoofax.terms/src/org/spoofax/terms/StrategoArrayList.java

```diff
@@ -1,289 +1,96 @@
-            }
-
-            return true;
-        }
-
-        final IStrategoList snd = (IStrategoList) second;
-
-        if(!isEmpty()) {
-            IStrategoTerm head = head();
-            IStrategoTerm head2 = snd.head();
-            if(head != head2 && !head.match(head2))
-                return false;
-
-            IStrategoList tail = tail();
-            IStrategoList tail2 = snd.tail();
-
-            for(IStrategoList cons = tail, cons2 = tail2; !cons.isEmpty(); cons = cons.tail(), cons2 = cons2.tail()) {
-                IStrategoTerm consHead = cons.head();
-                IStrategoTerm cons2Head = cons2.head();
-                if(!cons.getAnnotations().match(cons2.getAnnotations())) {
-                    return false;
-                }
-                if(consHead != cons2Head && !consHead.match(cons2Head))
-                    return false;
-            }
-        }
-
-        IStrategoList annotations = getAnnotations();
-        IStrategoList secondAnnotations = second.getAnnotations();
-        if(annotations == secondAnnotations) {
-            return true;
-        } else
-            return annotations.match(secondAnnotations);
-    }
-
-    /**
-     * N.B. this implementation may look strange but it's designed to return the same hashcode as {@link StrategoList#hashFunction()}
-     */
-    @Override protected int hashFunction() {
-        if(isEmpty())
-            return 1;
-
-        int i = offset;
-        int result = 31 * terms[i].hashCode();
-        for(i++; i < endOffset; i++) {
-            result = 31 * result + terms[i].hashCode();
-        }
-
-        return result;
-    }
-
-    @Override public Iterator<IStrategoTerm> iterator() {
-        return new StrategoArrayListIterator(this);
-    }
-
+    @SuppressWarnings("deprecation")
     @Override
-    public void putAttachment(ITermAttachment attachment) {
-        setTailAttachment(attachment);
-        super.putAttachment(attachment);
+    public void internalPrintAnnotations(ITermPrinter pp) {
+        printAnnotations(pp);
     }
 
     @Override
-    @Nullable
-    public ITermAttachment removeAttachment(TermAttachmentType<?> type) {
-        final ITermAttachment attachment = super.removeAttachment(type);
-        setTailAttachment(attachment());
-        return attachment;
+    protected boolean doSlowMatch(IStrategoTerm second) {
+        return internalDoSlowMatch(second);
     }
 
     @Override
-    protected void clearAttachments() {
-        setTailAttachment(null);
-        super.clearAttachments();
+    protected int hashFunction() {
+        return internalHashFunction();
     }
 
-    private void setTailAttachment(ITermAttachment attachment) {
-        if(offset > 0) {
-            tailAttachments[offset -1] = attachment;
+    @Override
+    public IStrategoList tail() {
+        if(isEmpty()) {
+            throw new IllegalStateException();
         }
+        return new StrategoArrayListTail(this, offset + 1);
     }
```

Comment:
Calling `tail()` on your list always returns a new object. I would argue that `tail()` is more often called than that attachments are set on the tails of lists. So, I would strongly favor, if possible, that setting the attachment on the tail of a list returns a new tail object with the attachment: no mutation in the list term. The fact that other terms can still be mutated should not matter, since there was never a guarantee that setting the attachment on an object returns the same object.

Reply:
Whether or not attachments are mutable is entirely orthogonal to this PR.